Flaky tests break CI intermittently

`tech.relaycorp.ping.ui.peers.AddPublicPeerActivityTest.addPublicPeerMissingCertificate` is the culprit: https://console.firebase.google.com/u/0/project/awala-ping-android/testlab/histories/bh.6e5c9f23e2ef2570/matrices/5757771105566649704/executions/bs.2400bf6e6bbc3d1/testcases/3/test-cases

```
com.schibsted.spain.barista.internal.failurehandler.BaristaException: No view matching (with string from resource id: <2131887517>[peer_add_missing_certificate] value: Missing Identity Certificate) was found
at com.schibsted.spain.barista.internal.failurehandler.SpyFailureHandler.resendFirstError(SpyFailureHandler.kt:25)
at com.schibsted.spain.barista.internal.AssertAnyKt.assertAnyView(AssertAny.kt:40)
at com.schibsted.spain.barista.internal.AssertAnyKt.assertAny(AssertAny.kt:18)
at com.schibsted.spain.barista.assertion.BaristaVisibilityAssertions.assertDisplayed(BaristaVisibilityAssertions.kt:33)
at tech.relaycorp.ping.ui.peers.AddPublicPeerActivityTest.addPublicPeerMissingCertificate(AddPublicPeerActivityTest.kt:71)
at java.lang.reflect.Method.invoke(Native Method)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at androidx.test.internal.runner.junit4.statement.RunBefores.evaluate(RunBefores.java:80)
at tech.relaycorp.ping.test.IntentsRule$apply$1.evaluate(IntentsRule.kt:18)
at com.schibsted.spain.barista.rule.cleardata.ClearFilesRule$apply$1.evaluate(ClearFilesRule.kt:25)
at tech.relaycorp.ping.test.ClearTestDatabaseRule$apply$1.evaluate(ClearTestDatabaseRule.kt:20)
at com.schibsted.spain.barista.rule.cleardata.ClearPreferencesRule$1.evaluate(ClearPreferencesRule.java:26)
at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:549)
at org.junit.rules.RunRules.evaluate(RunRules.java:20)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at androidx.test.ext.junit.runners.AndroidJUnit4.run(AndroidJUnit4.java:154)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:395)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2189)
Caused by: androidx.test.espresso.NoMatchingViewException: No views in hierarchy found matching: with string from resource id: <2131887517>[peer_add_missing_certificate] value: Missing Identity Certificate

View Hierarchy:
+>DecorView{id=-1, visibility=VISIBLE, width=1080, height=1920, has-focus=false, has-focusable=true, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params={(0,0)(fillxfill) sim={adjust=resize} ty=BASE_APPLICATION wanim=0x10302fe
fl=LAYOUT_IN_SCREEN LAYOUT_INSET_DECOR SPLIT_TOUCH HARDWARE_ACCELERATED DRAWS_SYSTEM_BAR_BACKGROUNDS
pfl=FORCE_DRAW_STATUS_BAR_BACKGROUND}, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, child-count=2}
|
+->LinearLayout{id=-1, visibility=VISIBLE, width=1080, height=1920, has-focus=false, has-focusable=true, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=android.widget.FrameLayout$LayoutParams@f702737, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, child-count=2}
|
+-->ViewStub{id=16908692, res-name=action_mode_bar_stub, visibility=GONE, width=0, height=0, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=true, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@a3da4a4, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0}
|
+-->FrameLayout{id=-1, visibility=VISIBLE, width=1080, height=1920, has-focus=false, has-focusable=true, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@3f745c2, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, child-count=1}
|
+--->FitWindowsLinearLayout{id=2131296319, res-name=action_bar_root, visibility=VISIBLE, width=1080, height=1920, has-focus=false, has-focusable=true, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=android.widget.FrameLayout$LayoutParams@bccf209, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, child-count=2}
|
+---->ViewStubCompat{id=2131296330, res-name=action_mode_bar_stub, visibility=GONE, width=0, height=0, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=true, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@3f7fc0e, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0}
|
+---->ContentFrameLayout{id=16908290, res-name=content, visibility=VISIBLE, width=1080, height=1920, has-focus=false, has-focusable=true, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@f0ee43c, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, child-count=1}
|
+----->CoordinatorLayout{id=-1, visibility=VISIBLE, width=1080, height=1920, has-focus=false, has-focusable=true, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=android.widget.FrameLayout$LayoutParams@549a5c5, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, child-count=2}
|
+------>AppBarLayout{id=2131296345, res-name=appBar, visibility=VISIBLE, width=1080, height=220, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=androidx.coordinatorlayout.widget.CoordinatorLayout$LayoutParams@d73964b, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, child-count=1}
|
+------->Toolbar{id=2131296729, res-name=toolbar, visibility=VISIBLE, width=1080, height=154, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=com.google.android.material.appbar.AppBarLayout$LayoutParams@2634c27, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=66.0, child-count=3}
|
+-------->MaterialTextView{id=2131296730, res-name=toolbarTitle, visibility=VISIBLE, width=390, height=69, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=androidx.appcompat.widget.Toolbar$LayoutParams@2a57ed4, tag=null, root-is-layout-requested=false, has-input-connection=false, x=345.0, y=42.0, text=Public Endpoint, input-type=0, ime-target=false, has-links=false}
|
+-------->AppCompatImageButton{id=-1, visibility=VISIBLE, width=154, height=154, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=true, is-enabled=true, is-focused=false, is-focusable=true, is-layout-requested=false, is-selected=false, layout-params=androidx.appcompat.widget.Toolbar$LayoutParams@913c87d, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0}
|
+-------->ActionMenuView{id=-1, visibility=VISIBLE, width=150, height=154, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=androidx.appcompat.widget.Toolbar$LayoutParams@cae83c3, tag=null, root-is-layout-requested=false, has-input-connection=false, x=930.0, y=0.0, child-count=1}
|
+--------->ActionMenuItemView{id=2131296620, res-name=save, visibility=VISIBLE, width=150, height=132, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=true, is-enabled=true, is-focused=false, is-focusable=true, is-layout-requested=false, is-selected=false, layout-params=androidx.appcompat.widget.ActionMenuView$LayoutParams@e45b440, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=11.0, text=Save, input-type=0, ime-target=false, has-links=false}
|
+------>LinearLayout{id=-1, visibility=VISIBLE, width=1080, height=455, has-focus=false, has-focusable=true, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=androidx.coordinatorlayout.widget.CoordinatorLayout$LayoutParams@b388a35, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=220.0, child-count=4}
|
+------->LinearLayout{id=-1, visibility=VISIBLE, width=1080, height=203, has-focus=false, has-focusable=true, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@75dcb58, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, child-count=2}
|
+-------->MaterialTextView{id=2, visibility=VISIBLE, width=309, height=203, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@4f494b1, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, text=Address *, input-type=0, ime-target=false, has-links=false}
|
+-------->TextInputEditText{id=2131296337, res-name=addressEdit, visibility=VISIBLE, width=727, height=120, has-focus=false, has-focusable=true, has-window-focus=true, is-clickable=true, is-enabled=true, is-focused=false, is-focusable=true, is-layout-requested=false, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@a8ccaed, tag=null, root-is-layout-requested=false, has-input-connection=true, editor-info=[inputType=0x11 imeOptions=0x6 privateImeOptions=null actionLabel=null actionId=0 initialSelStart=0 initialSelEnd=0 initialCapsMode=0x0 hintText=Type an address label=null packageName=null fieldId=0 fieldName=null extras=null hintLocales=null contentMimeTypes=null ], x=309.0, y=41.0, text=ping.awala.services, hint=Type an address, input-type=17, ime-target=false, has-links=false}
|
+------->View{id=-1, visibility=VISIBLE, width=1080, height=1, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@414cd22, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=203.0}
|
+------->LinearLayout{id=-1, visibility=VISIBLE, width=1080, height=250, has-focus=false, has-focusable=true, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@9f828e9, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=204.0, child-count=4}
|
+-------->MaterialTextView{id=-1, visibility=VISIBLE, width=309, height=250, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@856996e, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, text=Identity Certificate *, input-type=0, ime-target=false, has-links=false}
|
+-------->MaterialButton{id=2131296367, res-name=certificateButton, visibility=VISIBLE, width=727, height=132, has-focus=false, has-focusable=true, has-window-focus=true, is-clickable=true, is-enabled=true, is-focused=false, is-focusable=true, is-layout-requested=false, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@8b3080f, tag=null, root-is-layout-requested=false, has-input-connection=false, x=309.0, y=59.0, text=Upload certificate, input-type=0, ime-target=false, has-links=false, is-checked=false}
|
+-------->MaterialTextView{id=2131296369, res-name=certificateName, visibility=GONE, width=0, height=0, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=true, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@b49309c, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, text=, input-type=0, ime-target=false, has-links=false}
|
+-------->MaterialButton{id=2131296368, res-name=certificateClear, visibility=GONE, width=0, height=0, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=true, is-enabled=true, is-focused=false, is-focusable=true, is-layout-requested=true, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@fa16aa5, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=0.0, text=, input-type=0, ime-target=false, has-links=false, is-checked=false}
|
+------->View{id=-1, visibility=VISIBLE, width=1080, height=1, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@6d6de7a, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=454.0}
|
+->View{id=16908336, res-name=navigationBarBackground, visibility=VISIBLE, width=1080, height=132, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=android.widget.FrameLayout$LayoutParams@da2162b, tag=null, root-is-layout-requested=false, has-input-connection=false, x=0.0, y=1788.0}
|
at androidx.test.espresso.NoMatchingViewException$Builder.build(NoMatchingViewException.java:15)
at androidx.test.espresso.base.ViewFinderImpl.getView(ViewFinderImpl.java:33)
at androidx.test.espresso.ViewInteraction$2.call(ViewInteraction.java:4)
at androidx.test.espresso.ViewInteraction$2.call(ViewInteraction.java:2)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at android.os.Handler.handleCallback(Handler.java:883)
at android.os.Handler.dispatchMessage(Handler.java:100)
at android.os.Looper.loop(Looper.java:214)
at android.app.ActivityThread.main(ActivityThread.java:7356)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

` tech.relaycorp.ping.ui.peers.PeerActivityTest:deletes` has been failing often too: https://console.firebase.google.com/u/0/project/awala-ping-android/testlab/histories/bh.6e5c9f23e2ef2570/matrices/5124950595747160517/executions/bs.2b802cd05e29d925/testcases/5/errors

```
java.lang.AssertionError: Timeout waiting
at tech.relaycorp.ping.test.WaitAssertions.waitFor(WaitAssertions.kt:19)
at tech.relaycorp.ping.ui.peers.PeerActivityTest.deletes(PeerActivityTest.kt:64)
at java.lang.reflect.Method.invoke(Native Method)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at androidx.test.internal.runner.junit4.statement.RunBefores.evaluate(RunBefores.java:80)
at tech.relaycorp.ping.test.IntentsRule$apply$1.evaluate(IntentsRule.kt:18)
at com.schibsted.spain.barista.rule.cleardata.ClearFilesRule$apply$1.evaluate(ClearFilesRule.kt:25)
at tech.relaycorp.ping.test.ClearTestDatabaseRule$apply$1.evaluate(ClearTestDatabaseRule.kt:20)
at com.schibsted.spain.barista.rule.cleardata.ClearPreferencesRule$1.evaluate(ClearPreferencesRule.java:26)
at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:549)
at org.junit.rules.RunRules.evaluate(RunRules.java:20)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at androidx.test.ext.junit.runners.AndroidJUnit4.run(AndroidJUnit4.java:154)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:395)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2189)
Caused by: java.lang.AssertionError
at org.junit.Assert.fail(Assert.java:86)
at org.junit.Assert.assertTrue(Assert.java:41)
at org.junit.Assert.assertTrue(Assert.java:52)
at tech.relaycorp.ping.ui.peers.PeerActivityTest$deletes$2.invoke(PeerActivityTest.kt:65)
at tech.relaycorp.ping.ui.peers.PeerActivityTest$deletes$2.invoke(PeerActivityTest.kt:25)
at tech.relaycorp.ping.test.WaitAssertions.waitFor(WaitAssertions.kt:12)
... 36 more
```

--- a/.gitignore
+++ b/.gitignore
@@ -1,6 +1,7 @@
 # Built application files
 *.apk
 *.ap_
+app/release/
 
 # Files for the ART/Dalvik VM
 *.dex

--- a/app/src/androidTest/java/tech/relaycorp/ping/test/IntentsRule.kt
+++ b/app/src/androidTest/java/tech/relaycorp/ping/test/IntentsRule.kt
@@ -16,7 +16,11 @@ class IntentsRule : TestRule {
                     // Occasionally `Intents.init()` might be called twice before `Intents.release()` is called
                 }
                 base.evaluate()
-                Intents.release()
+                try {
+                    Intents.release()
+                } catch (_: IllegalStateException) {
+                    // Occasionally `Intents.init()` might be missed
+                }
             }
         }
 }

--- a/app/src/androidTest/java/tech/relaycorp/ping/ui/peers/AddPublicPeerActivityTest.kt
+++ b/app/src/androidTest/java/tech/relaycorp/ping/ui/peers/AddPublicPeerActivityTest.kt
@@ -10,9 +10,12 @@ import androidx.test.ext.junit.runners.AndroidJUnit4
 import com.schibsted.spain.barista.assertion.BaristaVisibilityAssertions.assertDisplayed
 import com.schibsted.spain.barista.interaction.BaristaClickInteractions.clickOn
 import com.schibsted.spain.barista.interaction.BaristaEditTextInteractions.writeTo
+import com.schibsted.spain.barista.rule.flaky.AllowFlaky
+import com.schibsted.spain.barista.rule.flaky.FlakyTestRule
 import org.junit.Before
 import org.junit.Rule
 import org.junit.Test
+import org.junit.rules.RuleChain
 import org.junit.runner.RunWith
 import tech.relaycorp.ping.R
 import tech.relaycorp.ping.test.ActivityAssertions.waitForCurrentActivityToBe
@@ -26,6 +29,11 @@ class AddPublicPeerActivityTest {
     @Rule
     @JvmField
     val testRule = BaseActivityTestRule(PeersActivity::class, false)
+
+    @Rule
+    @JvmField
+    val flakyChainRule = RuleChain.outerRule(FlakyTestRule())
+        .around(testRule)
 
     @Before
     fun setUp() {
@@ -59,6 +67,7 @@ class AddPublicPeerActivityTest {
     }
 
     @Test
+    @AllowFlaky(attempts = 3)
     fun addPublicPeerMissingCertificate() {
         testRule.start()
         clickOn(R.id.addPeer)

--- a/app/src/androidTest/java/tech/relaycorp/ping/ui/peers/PeerActivityTest.kt
+++ b/app/src/androidTest/java/tech/relaycorp/ping/ui/peers/PeerActivityTest.kt
@@ -3,12 +3,16 @@ package tech.relaycorp.ping.ui.peers
 import androidx.test.ext.junit.runners.AndroidJUnit4
 import com.schibsted.spain.barista.assertion.BaristaVisibilityAssertions.assertDisplayed
 import com.schibsted.spain.barista.interaction.BaristaClickInteractions.clickOn
+import com.schibsted.spain.barista.rule.flaky.AllowFlaky
+import com.schibsted.spain.barista.rule.flaky.FlakyTestRule
+import com.schibsted.spain.barista.rule.flaky.Repeat
 import kotlinx.coroutines.flow.first
 import kotlinx.coroutines.runBlocking
 import org.junit.Assert.assertTrue
 import org.junit.Before
 import org.junit.Rule
 import org.junit.Test
+import org.junit.rules.RuleChain
 import org.junit.runner.RunWith
 import tech.relaycorp.ping.R
 import tech.relaycorp.ping.data.database.dao.PublicPeerDao
@@ -21,12 +25,18 @@ import tech.relaycorp.ping.test.WaitAssertions.waitFor
 import tech.relaycorp.ping.ui.peers.PeerActivity
 import javax.inject.Inject
 
+
 @RunWith(AndroidJUnit4::class)
 class PeerActivityTest {
 
     @Rule
     @JvmField
     val testRule = BaseActivityTestRule(PeerActivity::class, false)
+
+    @Rule
+    @JvmField
+    val flakyChainRule = RuleChain.outerRule(FlakyTestRule())
+        .around(testRule)
 
     @Inject
     lateinit var publicPeerDao: PublicPeerDao
@@ -50,6 +60,7 @@ class PeerActivityTest {
     }
 
     @Test
+    @AllowFlaky(attempts = 3)
     fun deletes() {
         val peer = PublicPeerEntityFactory.build()
         runBlocking {

--- a/app/src/androidTest/java/tech/relaycorp/ping/ui/peers/PeersActivityTest.kt
+++ b/app/src/androidTest/java/tech/relaycorp/ping/ui/peers/PeersActivityTest.kt
@@ -2,6 +2,7 @@ package tech.relaycorp.ping.ui.peers
 
 import androidx.test.ext.junit.runners.AndroidJUnit4
 import com.schibsted.spain.barista.assertion.BaristaVisibilityAssertions.assertDisplayed
+import com.schibsted.spain.barista.rule.flaky.AllowFlaky
 import kotlinx.coroutines.runBlocking
 import org.junit.Before
 import org.junit.Rule
@@ -11,6 +12,7 @@ import tech.relaycorp.ping.data.database.dao.PublicPeerDao
 import tech.relaycorp.ping.test.AppTestProvider.component
 import tech.relaycorp.ping.test.BaseActivityTestRule
 import tech.relaycorp.ping.test.PublicPeerEntityFactory
+import tech.relaycorp.ping.test.WaitAssertions.waitFor
 import javax.inject.Inject
 
 @RunWith(AndroidJUnit4::class)
@@ -29,6 +31,7 @@ class PeersActivityTest {
     }
 
     @Test
+    @AllowFlaky(attempts = 3)
     fun showsPublicPeers() {
         val peer = PublicPeerEntityFactory.build()
         runBlocking {
@@ -37,6 +40,8 @@ class PeersActivityTest {
 
         testRule.start()
 
-        assertDisplayed(peer.publicAddress)
+        waitFor {
+            assertDisplayed(peer.publicAddress)
+        }
     }
 }

